Template injection detector

kyakdan opened the pull request.

kyakdan committed:
Add template injection detector

bertschneider commented on sanitizers/template.go:
```diff
@@ -0,0 +1,27 @@
+	} else {
+		GuideTowardsContainment(text, evilTemplateCommand, fakePC)
+	}
+	return tmpl.Parse(text)
```
I wonder if `Parse` is the correct method to hook. It shouldn't be an issue to create a "problematic" template, as long as it's not executed with `Execute`.

0xricksanchez replied:
IMO `Parse` is a valid hooking point to just check whether injection is possible at all. That said, I agree that we should hook `Execute` and `ExecuteTemplate` as well to find those cases where the tampered with template is actually "*used*".

kyakdan replied:
Good point! Parsing the `Parse` function may lead to false positives. I've changed the logic to hook `Execute` and `ExecuteTemplate` and check whether the template to be executed contains the injected action.

0xricksanchez commented on sanitizers/template.go:
```diff
@@ -0,0 +1,41 @@
+	} else {
+		GuideTowardsContainment(tmplText, evilTemplateCommand, hookId)
+	}
+}
```
LGTM. The SQL PR should be reworked first to e.g. apply changes that affect all PRs like the typo in `repot.go`.

Only little nitpicky thing I find a little ugly is that we now across als sanitizers have such differing but long function names for essentially the same functionality:

* `checkForEvilCommandAndGuideFuzzer` -> `exec.go`
* `isSyntaxError` + `GuideTowardsContainment` -> `sql.go`
* `checkForEvilActionAndGuideFuzzer` -> `template.go`

Furthermore, sometimes the `ReportFinding` call is within these helper functions like in the case of the template-sanitizer, whereas e.g. for the SQL-sanitizer the call to `ReportFinding` is directly embedded in each hook. Shouldn't we streamline this so future PRs have an easier time integrating new sanitizers as they just can follow a sound pattern that we introduce here at the beginning?

kyakdan replied:
I've refactored the detection and reporting logic so that all detectors now have to implement the `Detector` interface. This should make the overall logic easier to understand and unified among all sanitizers. Could you have a look?

0xricksanchez replied:
That looks a lot cleaner to me! I like the idea of having that common interface. With the exception of the nitpick this LGTM :)

kyakdan committed:
Refactor bug detection and reporting

Now each bug detector should implement the Detector interface. Bug
detectors return specialized errors representing specific types of bugs.
The detectors are then used in the hooks. This unifies the logic among
all sanitizers.

0xricksanchez commented on sanitizers/detectors/sql_injection.go:
```diff
@@ -0,0 +1,55 @@
+	regexp.MustCompile(`\S+ ERROR: syntax error at or near .* \(SQLSTATE 42601\)`),         // PostgreSQL error message
+}
+
+type SQLInjection struct {
```
nitpick: Only this injection technique got some comments on the class specific struct. I'd prefer either all of them getting some or none of them.

kyakdan replied:
Added

0xricksanchez commented:
PS: With having a common interface now this might be something we would want to document in a CONTRIBUTING.md
